### DISCUSSION TITLE
Ideas for v2

### DISCUSSION THREAD

**Discussion by ChillFish8** [poem-web/poem#279] [Ideas]
So I've noticed there is a v2 beta/alpha  👀 

We're using poem in quite a few big systems and noticed a couple of things that would be nice to change (in  my opinion) but would have been breaking changes:

- When matching a path based route i.e `/foo/:bar` with a path value that attempts to get the case into say an integer (`Path<u64>`) I think it would possibly be a better idea to have this return a 404 rather than a 400 status code, this tripped up some devs when linking it up with the frontend and during testing because they would think their parameters or body is invalid, not the fact that the URL is incorrect. 
Specifically, this gets confusing when you have routes that overlap i.e `/foo/bar` and `/foo/:id` because if you miss type `/foo/bar` with `/foo/bars` it'll go down the `/foo/:id` route and return a 400 if it can't convert the value. 
- Add a utility type similar to `Option<T>` along the lines of `NonRequired<T>` with varients `Some(T)` and `Missing` where the variant is only `Missing` if the field is not included at all. 
This came out of running into a time where we had editable fields but some fields were also nullable, which meant we couldn't just have them be Options or default and I think it's probably common enough to be worth being in some sort of utils pack?
- Including the name of the fields that errored out / did not pass validation when parsing a body or query parameter in the error. This would make creating more descriptive error messages considerably easier, especially when interacting with the frontend and we want to be able to go "field x" failed.

Those are really the only pain points I've run into overall in the some 20,000 LOC worth of API routes.
Obviously these are just my two cents but as aways this framework is great  ❤️

**Comment by sunli829** [comment]
Thank you so much for these suggestions! 🙂

> Add a utility type similar to Option<T> along the lines of NonRequired<T> with varients Some(T) and Missing where the variant is only Missing if the field is not included at all.

You can use `MaybeUndefined`

**Reply by ChillFish8** [reply]
Thats very true!
